vortex-array/src/compute/filter.rs

```diff
@@ -0,0 +1,78 @@
+use arrow_array::cast::AsArray;
+use vortex_dtype::{DType, Nullability};
+use vortex_error::VortexResult;
+
+use crate::arrow::FromArrowArray;
+use crate::{Array, ArrayDType, ArrayData, IntoArray, IntoCanonical};
+
+pub trait FilterFn {
+    /// Filter an array by the provided predicate.
+    fn filter(&self, predicate: &Array) -> Array;
+}
+
+/// Return a new array by applying a boolean predicate to select items from a base Array.
+///
+/// # Performance
+///
+/// This function attempts to amortize the cost of copying
+///
+/// # Panics
+///
+/// The `predicate` must receive an Array with type non-nullable bool, and will panic if this is
+/// not the case.
+pub fn filter(array: &Array, predicate: &Array) -> VortexResult<Array> {
+    assert_eq!(
+        predicate.dtype(),
+        &DType::Bool(Nullability::NonNullable),
+        "predicate must be non-nullable bool"
+    );
+    assert_eq!(
+        predicate.len(),
+        array.len(),
+        "predicate.len() must equal array.len()"
+    );
+
+    array.with_dyn(|a| {
+        if let Some(filter_fn) = a.filter() {
+            Ok(filter_fn.filter(array))
```

Comment:
Should this be `filter_fn.filter(predicate)`

Reply:
Whoops, yes

Reply:
Ill fix it in a follow up.